src/manifold/src/quickhull.cpp

```diff
@@ -85,34 +228,94 @@ void QuickHull::buildMesh(const VertexDataSource& pointCloud, bool CCW,
   // The planar case happens when all the points appear to lie on a two
   // dimensional subspace of R^3.
   planar = false;
-  createConvexHalfEdgeMesh();
+  createConvexHalfedgeMesh();
   if (planar) {
-    const size_t extraPointIndex = planarPointCloudTemp.size() - 1;
-    for (auto& he : mesh.halfEdges) {
-      if (he.endVertex == extraPointIndex) {
-        he.endVertex = 0;
+    const int extraPointIndex = planarPointCloudTemp.size() - 1;
+    for (auto& he : mesh.halfedges) {
+      if (he.endVert == extraPointIndex) {
+        he.endVert = 0;
       }
     }
     originalVertexData = pointCloud;
     planarPointCloudTemp.clear();
   }
-}
 
-ConvexHull QuickHull::getConvexHull(const VertexDataSource& pointCloud,
-                                    bool CCW, bool useOriginalIndices,
-                                    double epsilon) {
-  buildMesh(pointCloud, CCW, useOriginalIndices, epsilon);
-  return ConvexHull(mesh, originalVertexData, CCW, useOriginalIndices);
+  // reorder halfedges
+  Vec<Halfedge> halfedges(mesh.halfedges.size());
+  Vec<int> counts(mesh.halfedges.size(), 0);
+  Vec<int> mapping(mesh.halfedges.size());
+  Vec<int> faceMap(mesh.faces.size());
+
+  // reorder halfedges
+  // Since all faces are not used now (so we should start from 0 and just
+  // increment, we can later set the face id as index/3 for each halfedge
+  int j = 0;
+  for_each(
+      autoPolicy(mesh.halfedges.size()), countAt(0_uz),
+      countAt(mesh.halfedges.size()), [&](size_t i) {
+        if (mesh.halfedges[i].pairedHalfedge < 0) return;
+        if (mesh.faces[mesh.halfedges[i].face].isDisabled()) return;
+        if (AtomicAdd(counts[mesh.halfedges[i].face], 1) > 0) return;
+        int currIndex = AtomicAdd(j, 3);
+        mapping[i] = currIndex;
+        halfedges[currIndex + 0] = mesh.halfedges[i];
+
+        size_t k = mesh.halfedgeNext[i];
+        mapping[k] = currIndex + 1;
+        halfedges[currIndex + 1] = mesh.halfedges[k];
+
+        k = mesh.halfedgeNext[k];
+        mapping[k] = currIndex + 2;
+        halfedges[currIndex + 2] = mesh.halfedges[k];
+        halfedges[currIndex + 0].startVert = halfedges[currIndex + 2].endVert;
+        halfedges[currIndex + 1].startVert = halfedges[currIndex + 0].endVert;
+        halfedges[currIndex + 2].startVert = halfedges[currIndex + 1].endVert;
+      });
+  halfedges.resize(j);
+  // fix pairedHalfedge id
+  for_each(
+      autoPolicy(halfedges.size()), halfedges.begin(), halfedges.end(),
+      [&](Halfedge& he) { he.pairedHalfedge = mapping[he.pairedHalfedge]; });
+  counts.resize(pointCloud.size() + 1);
+  fill(counts.begin(), counts.end(), 0);
+
+  // remove unused vertices
+  for_each(autoPolicy(halfedges.size() / 3), countAt(0_uz),
+           countAt(halfedges.size() / 3), [&](size_t i) {
+             AtomicAdd(counts[halfedges[3 * i].startVert], 1);
```

Comment:
Food for thought: maybe atomic set will be better for performance, and we don't need the transform iterator later.